### ISSUE TITLE
False positive `access-member-before-definition` (E0203) if member is defined in parent class in different module

### ISSUE THREAD

**Issue by patrick96** [PyCQA/pylint#9009]
### Bug description

Directory structure:

```
pkg
├── __init__.py
├── mod1.py
└── mod2.py
```

`__init__.py`:
```python
from .mod1 import *
```

`mod1.py`:
```python
from . import mod2

class SubClass(mod2.SuperClass):
    def __init__(self):
        super().__init__()
        self.attr2 = None

    def meth1(self):
        if not self.attr1:
            return
        self.attr1 = "Something"
```

`mod2.py`:
```python
class SuperClass:
    def __init__(self):
        self.attr1 = None
```

Pylint produces both `access-member-before-definition` (`E0203`) and `attribute-defined-outside-init` for the accesses to `attr1` in `meth1`.

Clearly `attr1` is defined in the `SuperClass` constructor (which is called from `SubClass`).

I have noticed that the error and warning disappear for the following small changes:

* If the `__init__` method in `SubClass` is omitted
* If the import in `__init__.py` is removed or changed to importing only specific symbols (anything except `mod2` or `SubClass`) from `mod1` (such a symbol would have to be added to `mod1` first though).

### Command used

```shell
pylint --disable=C,R0903 pkg
```

### Pylint output

```shell
************* Module pkg.mod1
pkg/mod1.py:9:15: E0203: Access to member 'attr1' before its definition line 11 (access-member-before-definition)
pkg/mod1.py:11:8: W0201: Attribute 'attr1' defined outside __init__ (attribute-defined-outside-init)

------------------------------------------------------------------
Your code has been rated at 5.38/10 (previous run: 5.38/10, +0.00)
```

### Expected behavior

Pylint should not produce `E0203` or `W0201` for `attr1`.

### Pylint version

```shell
pylint 2.17.5
astroid 2.15.6
Python 3.11.5 (main, Sep  5 2023, 17:44:34) [GCC 12.3.0]
```

This also happens on `main` and I originally found this with pylint 2.4.4, astroid 2.3.3, and python 3.8.5

### OS / Environment

Ubuntu 23.04

### Additional dependencies

Not sure this is needed, but here goes:

```
$ pip freeze
astroid==2.15.6
dill==0.3.7
isort==5.12.0
lazy-object-proxy==1.9.0
mccabe==0.7.0
platformdirs==3.10.0
pylint==2.17.5
tomlkit==0.12.1
wrapt==1.15.0
```
